File: Graph/java/HasCycle.java
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class HasCycle {
  public static boolean detect(Map<String, List<String>> graph) {
    Set<String> visiting = new HashSet<>();
    Set<String> visited = new HashSet<>();

    for (String node : graph.keySet()) {
      if (cycle(graph, node, visiting, visited)) {
        return true;
      }
    }

    return false;
  }

  private static boolean cycle(Map<String, List<String>> graph, String node, Set<String> visiting, Set<String> visited) {
    if (visited.contains(node)) {
      return false;
    }
    if (visiting.contains(node)) {
      return true;
    }

    visiting.add(node);
    for (String neighbor : graph.getOrDefault(node, List.of())) {
      if (cycle(graph, neighbor, visiting, visited)) {
        return true;
      }
    }

    visiting.remove(node);
    visited.add(node);
    return false;
  }
}

File: Graph/python/hasCycle.py
from .graph_utils import Graph


def has_cycle(graph: Graph) -> bool:
    visiting: set[str] = set()
    visited: set[str] = set()

    for node in graph:
        if _detect(graph, node, visiting, visited):
            return True
    return False


def _detect(graph: Graph, node: str, visiting: set[str], visited: set[str]) -> bool:
    if node in visited:
        return False
    if node in visiting:
        return True

    visiting.add(node)
    for neighbor in graph.get(node, []):
        if _detect(graph, neighbor, visiting, visited):
            return True
    visiting.remove(node)
    visited.add(node)
    return False
